File: sdks/core/src/modules.ts
/**
 * The runtime dependency inventory that rides events: which package versions are actually loaded, as
 * opposed to which ones a manifest declares. The relay indexes this so a CVE advisory can be answered
 * with "and you are running 4.17.11 in production" instead of only "your lockfile says so".
 *
 * Collection is NOT here, deliberately. Enumerating installed packages needs a filesystem, and this
 * package is the isomorphic core: fetch, crypto and URL only, so it runs unchanged in a browser, an
 * edge worker and Node. @condux/node does the collecting and hands the result to `setModules`. A
 * browser or edge bundle has no installed tree to enumerate and sends nothing, which reads as unknown
 * on the server rather than as "no vulnerable version present".
 *
 * Kept apart from the ambient scope even though both ride events, because `clearScope()` means
 * "forget who the user is", and forgetting which packages are loaded at the same moment would be
 * surprising: they are a fact about the process, not about the session.
 */

/**
 * The most package entries carried on one event. A large Node tree runs past this, and the cap
 * applies after sorting, so which entries survive is stable across events rather than varying with
 * directory order: the server sees one consistent set instead of a shifting sample.
 */
export const MAX_MODULES = 1000;

/**
 * How long to wait before repeating the inventory on another event.
 *
 * <p><b>This is why the feature is affordable.</b> Measured on a real Next.js application, the
 * inventory is 451 packages and about 11KB of JSON, and a large monorepo root is double that.
 * Attaching it to every event, which is what the other SDKs in this space do, would add that to every
 * single error a busy project reports, for no gain: the server deduplicates a release's inventory
 * down to one row per package per day, so all but the first event's copy is discarded on arrival.</p>
 *
 * <p>Repeating on an interval rather than sending once keeps the robustness that the every-event
 * approach buys. The event carrying the inventory can be dropped by a rate limit or a quota rejection
 * before anything parses it, so a single attempt per process would lose that day's inventory
 * outright. At this interval a process gets many attempts a day and the server needs one.</p>
 */
export const MODULES_INTERVAL_MS = 15 * 60 * 1000;

let modules: Record<string, string> | undefined;
let lastAttachedAt: number | undefined;

/**
 * Declare the loaded package versions for subsequent events; null clears them. Sorted and capped
 * here so every caller gets the same treatment, and so a caller cannot accidentally make each event
 * carry a differently ordered map.
 */
export function setModules(next: Record<string, string> | null): void {
  // A fresh declaration is news, so let the next event carry it rather than waiting out an interval
  // started by the previous inventory.
  lastAttachedAt = undefined;

  if (next === null) {
    modules = undefined;
    return;
  }

  const entries = Object.entries(next)
    .filter(([name, version]) => name.length > 0 && typeof version === "string" && version.length > 0)
    .sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0))
    .slice(0, MAX_MODULES);
  modules = entries.length > 0 ? Object.fromEntries(entries) : undefined;
}

/**
 * The inventory's contribution to an event: the full map on the first event and then at most once per
 * {@link MODULES_INTERVAL_MS}, and nothing at all otherwise. Absent entirely when nothing was
 * declared, so an event from a browser keeps its exact previous wire shape.
 */
export function modulesField(now: number = Date.now()): { modules?: Record<string, string> } {
  if (modules === undefined) {
    return {};
  }
  if (lastAttachedAt !== undefined && now - lastAttachedAt < MODULES_INTERVAL_MS) {
    return {};
  }

  lastAttachedAt = now;
  return { modules: { ...modules } };
}

/** Reset the inventory and its interval. Tests only; a process has one installed tree. */
export function clearModules(): void {
  modules = undefined;
  lastAttachedAt = undefined;
}

File: sdks/go/modules.go
package condux

import (
	"runtime/debug"
	"sort"
	"sync"
	"time"
)

// The runtime dependency inventory that rides events (ADR-0041): which module versions are actually
// built into this binary, as opposed to which ones a go.mod declares. The relay indexes this so a
// security advisory can be answered with "and you are running v1.6.0 in production" rather than only
// "your go.mod says so".
//
// Go has the best source of this in the fleet, and it needs no filesystem walk and no guessing: the
// toolchain records the exact resolved version of every dependency INTO the binary at build time, and
// debug.ReadBuildInfo reads it back. Verified against a real third-party binary, which recorded 221
// dependencies with exact versions.

// maxModules is the most entries carried on one event. The cap applies after sorting, so which entries
// survive is stable across events rather than varying with map order: the server sees one consistent
// set instead of a shifting sample.
const maxModules = 1000

// modulesInterval is how long to wait before repeating the inventory on another event.
//
// This is what makes the feature affordable. The server deduplicates a release's inventory down to one
// row per package per day, so attaching the whole map to every event would spend bytes for nothing.
// Repeating on an interval rather than sending once keeps the robustness that every-event buys: the
// event carrying the inventory can be dropped by a rate limit or a quota rejection before anything
// parses it, so a single attempt per process would lose that day's inventory outright.
const modulesInterval = 15 * time.Minute

var (
	modulesMu       sync.Mutex
	modules         map[string]string
	lastAttachedAt  time.Time
	hasAttachedOnce bool
)

// collectModules reads the module versions recorded in this binary, empty when there are none.
//
// A binary built without module information, which is what `go build` on a GOPATH-style project or a
// heavily stripped build produces, reports nothing. Unknown is the honest answer there, and it is the
// same answer the server already understands from every SDK that cannot enumerate.
func collectModules() map[string]string {
	info, ok := debug.ReadBuildInfo()
	if !ok {
		return nil
	}

	found := make(map[string]string, len(info.Deps)+1)
	// The main module is the application itself. It is included when it carries a real version, and
	// skipped when it does not: an unreleased build reports "(devel)", which is not a version and would
	// render as "running (devel)" against an advisory it can never match.
	if isRealVersion(info.Main.Version) {
		found[info.Main.Path] = info.Main.Version
	}
	for _, dep := range info.Deps {
		if dep == nil {
			continue
		}
		// A replace directive means the built code is NOT the version the requirement names, so report
		// what was actually built. Reporting the replaced-away version would name a version this binary
		// does not contain, which is the one kind of wrong answer this feature must not give.
		effective := dep
		if dep.Replace != nil {
			effective = dep.Replace
		}
		if effective.Path != "" && isRealVersion(effective.Version) {
			// The requirement's module path is the name the ecosystem knows, even under a replace: OSV
			// indexes advisories against the required module, not against a local fork's path.
			found[dep.Path] = effective.Version
		}
	}
	return found
}

// isRealVersion rejects the placeholders the toolchain uses when there is no released version to name.
func isRealVersion(version string) bool {
	return version != "" && version != "(devel)"
}

// initModules declares the inventory a new client will report, or clears it when the client asked for
// none. Called once per client rather than per capture: a binary's build info is fixed, so there is
// nothing to re-read. Set explicitly either way, so constructing a client with the inventory disabled
// cannot inherit one a previous client declared, since this state is package-wide.
func initModules(disabled bool) {
	if disabled {
		setModules(nil)
		return
	}
	setModules(collectModules())
}

// setModules declares the module versions for subsequent events; nil clears them.
func setModules(next map[string]string) {
	modulesMu.Lock()
	defer modulesMu.Unlock()

	// A fresh declaration is news, so let the next event carry it rather than waiting out an interval
	// started by the previous inventory.
	hasAttachedOnce = false

	if len(next) == 0 {
		modules = nil
		return
	}

	names := make([]string, 0, len(next))
	for name, version := range next {
		if name != "" && version != "" {
			names = append(names, name)
		}
	}
	sort.Strings(names)
	if len(names) > maxModules {
		names = names[:maxModules]
	}

	trimmed := make(map[string]string, len(names))
	for _, name := range names {
		trimmed[name] = next[name]
	}
	if len(trimmed) == 0 {
		modules = nil
		return
	}
	modules = trimmed
}

// modulesField returns the inventory to attach to an event: the full map on the first event and then at
// most once per modulesInterval, and nil otherwise, so an event that carries nothing keeps its exact
// previous wire shape.
func modulesField(now time.Time) map[string]string {
	modulesMu.Lock()
	defer modulesMu.Unlock()

	if modules == nil {
		return nil
	}
	if hasAttachedOnce && now.Sub(lastAttachedAt) < modulesInterval {
		return nil
	}

	lastAttachedAt = now
	hasAttachedOnce = true
	return modules
}

// clearModules resets the inventory and its interval. Tests only; a binary has one build.
func clearModules() {
	modulesMu.Lock()
	defer modulesMu.Unlock()
	modules = nil
	lastAttachedAt = time.Time{}
	hasAttachedOnce = false
}
